plugins/cxxrandom.cpp

```diff
@@ -137,118 +123,100 @@ class NumberSequence
 public:
     NumberSequence(){}
-    NumberSequence( int64_t start, int64_t end )
-    {
-        for( int64_t i = start; i <= end; ++i )
-        {
+    NumberSequence( int64_t start, int64_t end ) {
+        for( int64_t i = start; i <= end; ++i ) {
             m_numbers.push_back( i );
         }
     }
     void Add( int64_t num ) { m_numbers.push_back( num ); }
-    void Reset()            { m_numbers.clear(); }
-    int64_t Next()
-    {
-        if(m_position >= m_numbers.size())
-        {
+    void Reset() { m_numbers.clear(); }
+    int64_t Next() {
+        if(m_position >= m_numbers.size()) {
             m_position = 0;
         }
         return m_numbers[m_position++];
     }
-    void Shuffle( uint16_t id )
-    {
-        std::shuffle( std::begin( m_numbers ), std::end( m_numbers ), EnginesKeeper::Instance().RNG( id ) );
+    void Shuffle( uint16_t engID ) {
+        std::shuffle( std::begin( m_numbers ), std::end( m_numbers ), EnginesKeeper::Instance().RNG(engID));
     }
-    void Print()
-    {
-        for( auto v : m_numbers )
-        {
+    void Print() {
+        for( auto v : m_numbers ) {
             cout->print( "%" PRId64 " ", v );
         }
     }
 };
 
+#define SK_ID_BASE 0
+
 class SequenceKeeper
 {
 private:
     SequenceKeeper() {}
     std::unordered_map<uint16_t, NumberSequence> m_sequences;
-    uint16_t counter = 0;
+    uint16_t id_counter = SK_ID_BASE;
 public:
-    static SequenceKeeper& Instance()
-    {
+    static SequenceKeeper& Instance() {
         static SequenceKeeper instance;
         return instance;
     }
-    uint16_t MakeNumSequence( int64_t start, int64_t end )
-    {
-        m_sequences[++counter] = NumberSequence( start, end );
-        return counter;
+    uint16_t MakeNumSequence( int64_t start, int64_t end ) {
+        m_sequences[++id_counter] = NumberSequence(start, end);
+        return id_counter;
     }
-    uint16_t MakeNumSequence()
-    {
-        m_sequences[++counter] = NumberSequence();
-        return counter;
+    uint16_t MakeNumSequence() {
+        m_sequences[++id_counter] = NumberSequence();
+        return id_counter;
     }
-    void DestroySequence( uint16_t id )
-    {
-        m_sequences.erase( id );
+    void DestroySequence( uint16_t seqID ) {
+        m_sequences.erase(seqID);
     }
-    void AddToSequence( uint16_t id, int64_t num )
-    {
-        CHECK_INVALID_ARGUMENT( m_sequences.find( id ) != m_sequences.end() );
-        m_sequences[id].Add( num );
+    void AddToSequence(uint16_t seqID, int64_t num ) {
+        CHECK_INVALID_ARGUMENT(m_sequences.find(seqID) != m_sequences.end());
+        m_sequences[seqID].Add(num);
     }
-    void Shuffle( uint16_t id, uint16_t rng_id )
-    {
-        CHECK_INVALID_ARGUMENT( m_sequences.find( id ) != m_sequences.end() );
-        m_sequences[id].Shuffle( rng_id );
+    void Shuffle(uint16_t seqID, uint16_t engID ) {
+        uint16_t sid = seqID >= SK_ID_BASE ? seqID : engID;
```

Comment:
This is fairly brittle. Since the ids are opaque to Lua, how about making them different types? Then you can figure out which is which by resolving the type sent to you from Lua.

Reply:
The types are sent to lua, that's the only way lua users can get an id. I could use different integer types, but I don't see how lua would retain such information. I think I'd be better off just uptyping to 32 or 64 bits and increasing the bins for the two id types. If it is a problem that is, as is they have acess to several thousand ids which granted isn't a tonne of they just leave df open for a few months and were to regularly use RNG

Reply:
I meant wrap the ids in a type that has an `_identity`. Then you can use `virtual_cast` to figure out what you have

Reply:
Sounds like a decent plan, no idea how to implement that though.

Reply:
Example: https://github.com/DFHack/dfhack/blob/8eb2831b7ed1b3062f9fde1b541da55b89bbf6ac/plugins/blueprint.cpp#L93